### ISSUE TITLE
Expand warm starts: writing_features_v2, PLS/RBF-SVR engines, and same-task algorithm ensembles

### ISSUE THREAD

**Issue by shmercer** [shmercer/pairwiseLLM#259]
## Motivation

The ELLIPSE Adaptive Pairing Study needs a prospectively defined extension of the warm-start workflow that separates:

1. **feature richness**: the existing compact `writing_features_v1` vs a broader `writing_features_v2`;
2. **algorithmic form**: elastic net vs PLS regression vs RBF-SVR;
3. **ensemble benefit**: equal-weight averaging of multiple algorithms trained on the **same task**.

This must be implemented generically in `pairwiseLLM`, not as study-specific code.

The current package already has strong foundations:

- `fit_warm_start_model()` implements nested CV, training-fold preprocessing, within-task standardized outcomes, OOF linear calibration, and untouched outer validation;
- `writing_features_v1` is a frozen 20-feature schema with a detailed feature inventory/audit;
- `ensemble_warm_start_models()` equal-weights calibrated standardized predictions, but its contract is explicitly for independently sourced **task models**;
- warm model validation/artifacts are currently glmnet-specific (`training$engine == "glmnet"`).

The new workflow must preserve all existing v1/glmnet behavior and add an explicit **same-task algorithm ensemble** rather than silently repurposing the current cross-task ensemble semantics.

---

# Scientific / API requirements

## 1. Preserve existing behavior

Backward compatibility is a hard requirement.

- Existing `writing_features_v1` schema, extraction, hashes, ordering, definitions, preprocessing, predictions, validation metrics, artifact loading, registry behavior, and examples must remain unchanged.
- Existing calls to `fit_warm_start_model()` with no new arguments must still use the current glmnet elastic-net workflow.
- Existing model formats 1/2 must continue to validate/load/predict.
- Existing `ensemble_warm_start_models()` semantics for independently sourced task models must remain unchanged.
- Do not reinterpret old artifacts under new semantics.
- Add regression tests that make v1/glmnet compatibility explicit.

## 2. Add `writing_features_v2`

Add a new additive schema rather than modifying v1.

### Selection rule

Use the **existing pinned feature inventory/audit as the source of truth**, not outcome performance.

`writing_features_v2` must contain:

- every v1 feature, in the exact existing v1 order first;
- additional English, document-level **scalar** features already exposed by the pinned TextDescriptives/textstat stack when they are not excluded by the rules below.

Exclude:

- exact aliases;
- deterministic algebraic/algebraic-conditional restatements of retained features;
- utilities, predicates, lists, or other non-scalar outputs;
- non-English measures;
- corpus/web **quality-filtering** features and boilerplate/format heuristics;
- alternative fine-grained POS/tag configurations;
- fields already identified in the audit as primarily prompt/topic/format/annotation diagnostics (e.g. INTJ, NUM, PROPN, PUNCT, SYM, X composition).

Include otherwise eligible conceptual-overlap features intentionally: v2 is the **expanded** representation and the downstream regularized/latent/nonlinear algorithms may use correlated information.

Closed-class grammatical additions such as AUX, DET, and PART are eligible.

### Reproducibility

Before implementation, derive and commit an explicit audited v2 schema/inventory from the rule above. The schema must be deterministic and reviewable.

Requirements:

- stable feature order: all v1 fields first, then new fields in a documented deterministic inventory order;
- schema identifier exactly `writing_features_v2`;
- schema file SHA exposed/auditable as v1 is;
- no new Python package is required merely to create v2 unless absolutely necessary; prefer the already audited spaCy + TextDescriptives + textstat stack;
- extraction remains one row per item with `item_id` + exact schema features;
- preserve upstream missing values rather than silently inventing values; model preprocessing handles permitted missingness exactly as designed;
- audit every feature definition/version/source just as thoroughly as v1.

## 3. Generalize warm-start fitting to multiple engines

Extend the public warm-start training API to support at least:

- `engine = "glmnet"` — existing elastic net;
- `engine = "pls"` — partial least squares regression;
- `engine = "svr_rbf"` — radial-basis support-vector regression.

Preferred public shape:

```r
fit_warm_start_model(
  ids,
  theta,
  task_id,
  ...,
  engine = c("glmnet", "pls", "svr_rbf"),
  cv_plan = NULL,
  engine_control = NULL
)
```

Exact argument names may change if a cleaner API is justified, but the behavior below is required.

### Shared rules across engines

All engines must use the same statistical safeguards already established by the glmnet implementation:

- within-task standardized BTL outcome;
- preprocessing learned **inside each training split**;
- no preprocessing leakage from held-out rows;
- nested outer/inner validation;
- untouched outer predictions define validation performance;
- OOF linear calibration learned only from training-side OOF predictions;
- finite negative calibration slopes remain legal;
- final deployment model trained only after validation;
- deterministic seeds/folds;
- portable/auditable artifacts;
- no implicit dependency installation;
- raw training text never stored.

### Explicit reusable CV plan

Add a portable CV-plan object/API so multiple algorithms can use **exactly the same outer and inner partitions**.

For example:

```r
plan <- make_warm_start_cv_plan(
  ids,
  theta,
  seed = 123,
  outer_folds = 5,
  inner_folds = 5
)

en  <- fit_warm_start_model(..., engine = "glmnet",  cv_plan = plan)
pls <- fit_warm_start_model(..., engine = "pls",     cv_plan = plan)
svr <- fit_warm_start_model(..., engine = "svr_rbf", cv_plan = plan)
```

The plan must:

- bind item IDs/outcome identity sufficiently to prevent accidental reuse on a different task;
- preserve current outcome-balanced fold construction;
- record outer fold IDs and the inner fold plans needed by every outer/full tuning stage;
- be validated before fitting;
- be stored/referenced in audit metadata;
- ensure engine-specific RNG consumption cannot change the evaluation partitions.

## 4. PLS engine

Suggested optional dependency: `pls` in Suggests.

Use regression PLS with preprocessing from the existing warm-start preprocessing layer.

Prospective tuning contract:

- candidate component counts are positive integers;
- maximum components should be bounded by the available training rank/predictor count and **capped at 10**;
- derive a candidate set that is legal across all inner fits rather than silently dropping impossible candidates;
- tune by the same observation-count-weighted inner-fold MSE framework;
- use a **1-SE rule favoring fewer components**;
- deterministic tie handling;
- record the complete tuning trace and selected component count;
- final portable artifact should store only the numeric information required for prediction, not an opaque fitted object if that would violate the package's portability contract.

Because PLS deployment is ultimately linear in the processed predictors, prefer storing an equivalent numeric coefficient/intercept representation when practical.

## 5. RBF-SVR engine

Suggested optional dependency: `e1071` in Suggests, unless another lightweight CRAN implementation has a clearly stronger portability rationale.

Inputs are the existing training-fold standardized predictors and standardized outcome.

Prospective default tuning grid:

- `cost = 2^(-2:4)`;
- `gamma_multiplier = 2^(-2:2)`;
- actual `gamma = gamma_multiplier / p_retained` within the training fit;
- `epsilon = 0.10` fixed for the primary default.

Tune by observation-count-weighted inner-fold MSE.

Tie / 1-SE behavior should prefer the more regularized/smoother configuration:

1. lower cost;
2. lower gamma multiplier.

Record the complete tuning trace.

The saved warm model must remain portable and auditable. Do **not** require serializing an opaque external-pointer fit for deployment. Prefer a pure-R/numeric prediction payload containing the support vectors, dual coefficients, intercept/rho, gamma, and any other quantities required to reproduce predictions exactly.

Add round-trip tests comparing package prediction to the engine fit on held-out synthetic data.

## 6. Engine-neutral model artifact

The current validator assumes glmnet coefficients and `training$engine == "glmnet"`.

Refactor carefully.

Requirements:

- old formats 1/2 continue to work unchanged;
- introduce a new format/version or engine payload cleanly rather than weakening validation of old objects;
- every artifact records:
  - engine name/version;
  - schema;
  - preprocessing;
  - task/outcome definition;
  - hyperparameters;
  - tuning trace;
  - common CV-plan identity;
  - OOF calibration;
  - outer validation predictions/metrics;
  - engine-specific portable prediction payload;
- `predict()`, `summary()`, save/load, prepare/reduce, registry listing, metadata, and coefficient/introspection helpers should behave sensibly for each engine;
- if `warm_start_coefficients()` is not meaningful for nonlinear SVR, return a deliberate typed/explicit result or informative error rather than pretending coefficients exist.

## 7. Same-task algorithm ensemble

Do **not** change the semantics of `ensemble_warm_start_models()`.

Add a distinct API, e.g.:

```r
ensemble_warm_start_algorithms(
  elastic_net = en,
  pls = pls,
  svr_rbf = svr
)
```

This ensemble is for algorithms fit to the **same task / same rows / same standardized outcome / same feature schema / same CV plan**.

Hard requirements:

- at least two components;
- components must share task identity, schema, item IDs, outcome definition, and CV plan;
- reject cross-task components;
- reject incompatible schemas/folds/IDs;
- each component must have `oof_linear` calibration and complete outer validation;
- equal weights only for now;
- no learned stacking weights;
- no ensemble-level recalibration.

### Deployment prediction

For each item:

```
ensemble_prediction =
  mean(component calibrated_prediction)
```

The mean is taken after each component's own learned OOF calibration.

Return component predictions and between-algorithm SD as diagnostics if useful, but that SD is **not** a Bayesian prior SD.

### Honest ensemble validation

Ensemble validation must be computed from the aligned **outer-held-out calibrated predictions** of the component models:

```
outer_ensemble_prediction_i =
  mean(component outer-held-out calibrated predictions for i)
```

Then compute the standard validation metrics against the already aligned held-out observed standardized outcome.

Do not report an ensemble validation statistic constructed from in-sample/full-model predictions.

## 8. Ensemble artifact / prediction interoperability

The same-task algorithm ensemble must work wherever a warm predictor artifact is expected:

- `predict()`;
- save/load;
- prepare/reduce artifact;
- registry;
- summary/print;
- generation of warm-start predictions;
- `make_warm_start_prior()`;
- adaptive BTL and TrueSkill initialization downstream.

The ensemble's calibrated mean prediction is the score that becomes the warm prior; prior SD remains controlled by the adaptive prior constructor, not model disagreement.

## 9. Documentation

Update at minimum:

- warm-start model documentation;
- feature schema documentation;
- Python/environment audit documentation;
- ensemble documentation;
- README/model overview if it enumerates supported warm-start capabilities;
- relevant vignette(s), with an end-to-end example:
  - extract v2;
  - construct one shared CV plan;
  - fit glmnet / PLS / RBF-SVR;
  - inspect fair nested-CV metrics;
  - build same-task equal-weight ensemble;
  - predict new essays;
  - create warm prior;
  - use in adaptive ranking.

Clearly distinguish:
- cross-task ensemble (`ensemble_warm_start_models()`);
- same-task algorithm ensemble (new API).

## 10. Dependencies / CRAN

- Add optional engine packages to Suggests, not Imports, unless there is a compelling reason otherwise.
- Guard optional engines with informative dependency checks.
- No installation/download side effects.
- Keep examples CRAN-safe and lightweight.
- Review package check implications and conditional tests.
- Preserve current minimum R version unless a new dependency forces a justified change.

---

# Test requirements

Please add focused tests for all of the following.

### Regression/backward compatibility

- exact v1 schema/order unchanged;
- existing glmnet default path unchanged;
- old saved formats 1/2 validate/load/predict;
- existing cross-task ensemble tests continue passing.

### v2 features

- exact v2 schema/order and deterministic schema hash;
- v2 is a strict additive superset of v1;
- deterministic extraction;
- correct missing-value preservation;
- audit inventory covers every v2 field;
- feature names/types/order validated.

### CV plan

- deterministic construction under seed;
- same plan reused across engines;
- mismatch in IDs/outcomes/task rejected;
- no engine RNG can mutate folds;
- serialization/round trip.

### Engine fitting

For glmnet, PLS, and RBF-SVR:

- nested CV has no preprocessing leakage;
- untouched outer validation only;
- deterministic tuning/ties;
- legal edge cases for small p / reduced retained features;
- OOF calibration uses only training-side OOF predictions;
- final prediction finite;
- save/load prediction identity.

### PLS

- component grid legality;
- 1-SE favors fewer components;
- portable coefficient prediction matches fitted engine.

### SVR

- fixed default cost/gamma-multiplier grid;
- gamma scaling by retained p;
- regularization tie rules;
- portable numeric payload reproduces reference SVR predictions.

### Same-task ensemble

- requires same task/schema/IDs/CV plan;
- equal-weight calibrated deployment prediction;
- outer-held-out ensemble validation;
- component order/names stable;
- nested ensembles rejected unless explicitly designed otherwise;
- disagreement SD marked diagnostic only;
- prior creation works from ensemble predictions.

### Artifact/registry

- full and reduced artifacts;
- model and algorithm-ensemble save/load;
- listing reports artifact type, component algorithms, engine metadata, schema, validation;
- corruption/incompatibility failures remain explicit.

---

# Study-specific acceptance context (do not hard-code into package)

The immediate downstream study will use:

- multi-prompt training set only for this extension;
- N = 80 labels per trait;
- six traits;
- common 5 outer / 5 inner fold CV plan;
- `writing_features_v2`;
- three algorithms:
  - elastic net,
  - PLS,
  - RBF-SVR;
- equal-weight algorithm ensemble;
- no learned ensemble weights.

The study will compare nominal predictor tiers:

1. `compact_en` = existing v1 elastic net;
2. `expanded_en` = v2 elastic net;
3. `expanded_ensemble` = v2 equal-weight EN + PLS + RBF-SVR.

PLS and SVR component models will be retained/evaluated for audit and descriptive model-quality purposes but will **not** each become additional adaptive-replay factors.

The package should remain fully generic; these study names/tier choices belong in the study repository, not package APIs.

---

# Suggested implementation order / Codex tasklist

## Phase 1 — architecture and compatibility
- [ ] Read all warm-start implementation/tests/docs before editing.
- [ ] Add design notes for engine-neutral artifacts and same-task ensemble semantics.
- [ ] Add CV-plan object + validators.
- [ ] Refactor internals so current glmnet behavior runs through the generalized interface without changing results.
- [ ] Add explicit backward-compatibility regression tests.

## Phase 2 — feature schema v2
- [ ] Materialize audited v2 feature inventory from the frozen selection rule.
- [ ] Add schema resource + exact validation/hash tests.
- [ ] Extend Python extraction for every v2 field.
- [ ] Update Python environment/audit tooling.
- [ ] Add missingness and deterministic extraction tests.

## Phase 3 — PLS
- [ ] Add optional `pls` dependency guard.
- [ ] Implement legal common component grid.
- [ ] Implement nested tuning + 1-SE rule.
- [ ] Store portable linear deployment representation.
- [ ] Add engine-specific tests.

## Phase 4 — RBF-SVR
- [ ] Add optional `e1071` dependency guard.
- [ ] Implement frozen default grid and tie rules.
- [ ] Implement portable numeric RBF-SVR prediction payload.
- [ ] Verify prediction equality to reference engine.
- [ ] Add engine-specific tests.

## Phase 5 — same-task ensemble
- [ ] Add new algorithm-ensemble class/API without changing cross-task ensemble semantics.
- [ ] Validate shared CV plan/task/schema/IDs/outcomes.
- [ ] Implement equal-weight calibrated deployment predictions.
- [ ] Implement honest outer-held-out ensemble validation.
- [ ] Add save/load/registry/prior interoperability.
- [ ] Add complete tests.

## Phase 6 — documentation / package checks
- [ ] Update roxygen and regenerate NAMESPACE/Rd.
- [ ] Update vignettes/examples.
- [ ] Update README/model docs as needed.
- [ ] Run targeted tests.
- [ ] Run full `devtools::test()`.
- [ ] Run `R CMD check --as-cran` / equivalent and resolve new NOTES/WARNINGs attributable to this change.
- [ ] Review test coverage for all new R files/branches.
- [ ] Record final dependency versions and schema hashes.

---

# Non-goals

Do not add in this issue:

- learned stacking weights;
- random forest/XGBoost/neural-network engines;
- outcome-based feature screening;
- production-study-specific condition names;
- automatic model/source selection based on validation performance;
- changing Bayesian prior SD from model disagreement;
- changing adaptive pairing logic.

Those can be separate future work if justified.

---

# Definition of done

This issue is complete when:

1. v1/glmnet behavior is demonstrably backward compatible;
2. `writing_features_v2` is explicit, audited, deterministic, and documented;
3. glmnet, PLS, and RBF-SVR share one leakage-safe nested-CV framework and reusable fold plan;
4. all engines produce portable calibrated warm-model artifacts;
5. a distinct same-task equal-weight algorithm ensemble has honest outer-CV validation and deployment prediction;
6. ensemble/model predictions flow into the existing warm-prior/adaptive APIs;
7. unit/integration/regression tests pass;
8. package documentation and vignettes explain the workflows and semantics;
9. CRAN-style package checks are clean apart from previously known/justified conditions;
10. no existing cross-task ensemble or v1 artifact semantics have been silently changed.

**Comment by shmercer** [comment]
Study governance cross-reference: this package work is now frozen prospectively in the ELLIPSE Adaptive Pairing Study as D041 (`docs/decisions/D041-add-expanded-feature-multialgorithm-warm-start-extension.md` in `shmercer/ellipse-adaptive-pairing`).

Important timing/guardrail: D041 was adopted after compact-model D037 prior-quality summaries were observed but before any production adaptive replay. Therefore issue #259 must be implemented strictly from the outcome-independent feature/algorithm/tuning rules in this issue/D041; observed D037 values must not be used to curate v2 features, grids, algorithms, ensemble weights, or trait-specific exceptions.

The study will not fit substantive v2 models from an unpinned development checkout. After this issue is complete, the study will create D042 to freeze the exact pairwiseLLM version/commit and revalidate v1 compatibility, v2 schema identity, common CV plans, all three engine artifacts, same-task ensemble behavior, and warm-prior interoperability before any v2 fitting or D040 replay-harness execution.
